File: src/lib/discover.ts
/**
 * Route discovery: finds signposted cycle routes from OpenStreetMap (Overpass)
 * and generates loop rides with a cycling router. All client side — no keys.
 */
import { haversine, type RidePoint } from "./gpx";
import { type LatLon } from "./rejoin";

const OVERPASS_ENDPOINTS = [
  "https://overpass-api.de/api/interpreter",
  "https://overpass.kumi.systems/api/interpreter",
];
const OSRM_BIKE = "https://routing.openstreetmap.de/routed-bike/route/v1/bike";

export type DiscoveredRoute = {
  id: string;
  name: string;
  kind: "osm" | "loop";
  /** e.g. "Regional network" or "Generated loop" */
  subtitle: string;
  path: LatLon[];
  distanceM: number;
};

export function pathLengthM(path: LatLon[]): number {
  let total = 0;
  for (let i = 1; i < path.length; i += 1) {
    total += haversine(path[i - 1].lat, path[i - 1].lon, path[i].lat, path[i].lon);
  }
  return total;
}

/** Convert a discovered path into ride points (flat elevation — OSM has none). */
export function toRidePoints(path: LatLon[]): RidePoint[] {
  let d = 0;
  return path.map((point, index) => {
    if (index > 0) {
      d += haversine(path[index - 1].lat, path[index - 1].lon, point.lat, point.lon);
    }
    return { lat: point.lat, lon: point.lon, ele: 0, d };
  });
}

export function boundsOf(path: LatLon[]) {
  const lats = path.map((p) => p.lat);
  const lons = path.map((p) => p.lon);
  return {
    minLat: Math.min(...lats),
    minLon: Math.min(...lons),
    maxLat: Math.max(...lats),
    maxLon: Math.max(...lons),
  };
}

const NETWORK_LABEL: Record<string, string> = {
  icn: "International network",
  ncn: "National network",
  rcn: "Regional network",
  lcn: "Local network",
};

type OverpassWay = {
  type: string;
  id: number;
  geometry?: { lat: number; lon: number }[];
};

type OverpassRelation = {
  type: "relation";
  id: number;
  tags?: Record<string, string>;
  members?: { type: string; ref: number; geometry?: { lat: number; lon: number }[] }[];
};

/** Stitch relation member ways into one continuous-ish polyline. */
function stitch(members: OverpassRelation["members"]): LatLon[] {
  const segments = (members ?? [])
    .filter((member) => member.type === "way" && (member.geometry?.length ?? 0) > 1)
    .map((member) => member.geometry!.map((g) => ({ lat: g.lat, lon: g.lon })));
  if (segments.length === 0) return [];

  const remaining = segments.slice(1);
  const path = [...segments[0]];
  while (remaining.length > 0) {
    const tail = path[path.length - 1];
    let bestIndex = 0;
    let bestDistance = Infinity;
    let flip = false;
    remaining.forEach((segment, index) => {
      const start = haversine(tail.lat, tail.lon, segment[0].lat, segment[0].lon);
      const end = haversine(
        tail.lat,
        tail.lon,
        segment[segment.length - 1].lat,
        segment[segment.length - 1].lon,
      );
      const distance = Math.min(start, end);
      if (distance < bestDistance) {
        bestDistance = distance;
        bestIndex = index;
        flip = end < start;
      }
    });
    // A gap larger than 2 km means the rest belongs to a detached branch.
    if (bestDistance > 2000) break;
    const next = remaining.splice(bestIndex, 1)[0];
    path.push(...(flip ? [...next].reverse() : next));
  }
  return path;
}

async function overpass(query: string, signal: AbortSignal): Promise<any> {
  let lastError: unknown = new Error("Overpass unavailable");
  for (const endpoint of OVERPASS_ENDPOINTS) {
    try {
      const response = await fetch(endpoint, {
        method: "POST",
        body: new URLSearchParams({ data: query }),
        signal,
      });
      if (!response.ok) throw new Error(`Overpass ${response.status}`);
      return await response.json();
    } catch (error) {
      if (signal.aborted) throw error;
      lastError = error;
    }
  }
  throw lastError;
}

/** Signposted cycle routes whose ways pass near the given centre. */
export async function findNearbyRoutes(
  center: LatLon,
  radiusM: number,
  signal: AbortSignal,
): Promise<DiscoveredRoute[]> {
  const radius = Math.round(Math.min(Math.max(radiusM, 2000), 40000));
  const query = `[out:json][timeout:40];
relation["route"="bicycle"]["name"](around:${radius},${center.lat.toFixed(5)},${center.lon.toFixed(5)});
out geom 40;`;
  const data = await overpass(query, signal);
  const elements: (OverpassRelation | OverpassWay)[] = data?.elements ?? [];

  const routes: DiscoveredRoute[] = [];
  for (const element of elements) {
    if (element.type !== "relation") continue;
    const relation = element as OverpassRelation;
    const path = stitch(relation.members);
    if (path.length < 2) continue;
    const distanceM = pathLengthM(path);
    // Skip tiny connector segments — they aren't rides.
    if (distanceM < 5000) continue;
    const tags = relation.tags ?? {};
    const network = tags["network"] ?? "";
    routes.push({
      id: `osm-${relation.id}`,
      name: tags["name"] ?? "Unnamed cycle route",
      kind: "osm",
      subtitle:
        NETWORK_LABEL[network] ??
        (tags["operator"] ? tags["operator"] : "Signposted cycle route"),
      path,
      distanceM,
    });
  }
  return routes.sort((a, b) => b.distanceM - a.distanceM).slice(0, 12);
}

function offset(center: LatLon, bearingDeg: number, distanceM: number): LatLon {
  const R = 6371000;
  const bearing = (bearingDeg * Math.PI) / 180;
  const lat1 = (center.lat * Math.PI) / 180;
  const lon1 = (center.lon * Math.PI) / 180;
  const angular = distanceM / R;
  const lat2 = Math.asin(
    Math.sin(lat1) * Math.cos(angular) + Math.cos(lat1) * Math.sin(angular) * Math.cos(bearing),
  );
  const lon2 =
    lon1 +
    Math.atan2(
      Math.sin(bearing) * Math.sin(angular) * Math.cos(lat1),
      Math.cos(angular) - Math.sin(lat1) * Math.sin(lat2),
    );
  return { lat: (lat2 * 180) / Math.PI, lon: (lon2 * 180) / Math.PI };
}

async function routeVia(points: LatLon[], signal: AbortSignal) {
  const coords = points.map((p) => `${p.lon},${p.lat}`).join(";");
  const response = await fetch(`${OSRM_BIKE}/${coords}?overview=full&geometries=geojson`, {
    signal,
  });
  if (!response.ok) throw new Error(`OSRM ${response.status}`);
  const data = await response.json();
  const geometry = data?.routes?.[0]?.geometry?.coordinates;
  if (!Array.isArray(geometry) || geometry.length < 2) throw new Error("No route geometry");
  return {
    path: geometry.map(([lon, lat]: [number, number]) => ({ lat, lon })),
    distanceM: Number(data.routes[0].distance) || 0,
  };
}

/**
 * Loop rides of roughly the requested length, starting and finishing at the
 * given point. Each loop heads out on a different bearing.
 */
export async function generateLoops(
  start: LatLon,
  targetM: number,
  signal: AbortSignal,
  count = 3,
): Promise<DiscoveredRoute[]> {
  // Triangle-ish loop: three waypoints on a circle sized so the ridden
  // distance lands near the target.
  const radius = (targetM / (2 * Math.PI)) * 0.9;
  const bearings = Array.from({ length: count }, (_, index) => (360 / count) * index + 20);

  const results = await Promise.allSettled(
    bearings.map(async (bearing, index) => {
      const waypoints = [
        start,
        offset(start, bearing, radius),
        offset(start, bearing + 120, radius),
        offset(start, bearing + 240, radius),
        start,
      ];
      const routed = await routeVia(waypoints, signal);
      return {
        id: `loop-${index}`,
        name: `${Math.round(routed.distanceM / 1000)} km loop · ${compass(bearing)}`,
        kind: "loop" as const,
        subtitle: "Generated loop from your start point",
        path: routed.path,
        distanceM: routed.distanceM,
      };
    }),
  );

  const loops: DiscoveredRoute[] = [];
  for (const result of results) {
    if (result.status === "fulfilled") loops.push(result.value);
  }
  return loops;
}

function compass(bearingDeg: number): string {
  const names = ["north", "north-east", "east", "south-east", "south", "south-west", "west", "north-west"];
  return names[Math.round(((bearingDeg % 360) / 45)) % 8];
}

File: android-native/app/src/main/java/app/hodora/mobile/discover/Discover.kt
package app.hodora.mobile.discover

import app.hodora.mobile.gpx.computeAscentDescent
import app.hodora.mobile.gpx.haversine
import app.hodora.mobile.gpx.toDeg
import app.hodora.mobile.gpx.toRad
import app.hodora.mobile.net.HttpClientProvider
import app.hodora.mobile.routing.BikeProfile
import app.hodora.mobile.routing.LatLon
import app.hodora.mobile.routing.fetchRoute
import app.hodora.mobile.routing.pathLengthM
import app.hodora.mobile.routing.toRidePoints
import io.ktor.client.request.forms.submitForm
import io.ktor.client.statement.bodyAsText
import io.ktor.http.Parameters
import io.ktor.http.isSuccess
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

/**
 * Direct port of src/lib/discover.ts: finds signposted cycle routes from
 * OpenStreetMap (Overpass) and generates loop rides with a cycling router.
 * All client-side — no keys. boundsOf/toRidePoints are already ported (see
 * routing/Routing.kt's doc comments — they were pulled over early since the
 * planner needed the same shapes); only the Overpass lookup and loop
 * generator are new here.
 */
enum class DiscoveredRouteKind { OSM, LOOP }

data class DiscoveredRoute(
    val id: String,
    val name: String,
    val kind: DiscoveredRouteKind,
    /** e.g. "Regional network" or "Generated loop from your start point" */
    val subtitle: String,
    val path: List<LatLon>,
    val distanceM: Double,
    /** total climbing in meters, when the source had real elevation (generated loops via BRouter) — 0 for signposted OSM routes, which have none */
    val ascentM: Double,
)

private const val OVERPASS_PRIMARY = "https://overpass-api.de/api/interpreter"
private const val OVERPASS_FALLBACK = "https://overpass.kumi.systems/api/interpreter"

private val NETWORK_LABEL =
    mapOf(
        "icn" to "International network",
        "ncn" to "National network",
        "rcn" to "Regional network",
        "lcn" to "Local network",
    )

private val json = Json { ignoreUnknownKeys = true }

@Serializable
private data class OverpassResponse(val elements: List<OverpassElement> = emptyList())

@Serializable
private data class OverpassElement(
    val type: String,
    val id: Long,
    val tags: Map<String, String>? = null,
    val members: List<OverpassMember>? = null,
)

@Serializable
private data class OverpassMember(val type: String, val geometry: List<OverpassLatLon>? = null)

@Serializable
private data class OverpassLatLon(val lat: Double, val lon: Double)

/** Stitch relation member ways into one continuous-ish polyline. */
private fun stitch(members: List<OverpassMember>?): List<LatLon> {
    val segments =
        (members ?: emptyList())
            .filter { it.type == "way" && (it.geometry?.size ?: 0) > 1 }
            .map { member -> member.geometry!!.map { LatLon(it.lat, it.lon) } }
    if (segments.isEmpty()) return emptyList()

    val remaining = segments.drop(1).toMutableList()
    val path = segments[0].toMutableList()
    while (remaining.isNotEmpty()) {
        val tail = path.last()
        var bestIndex = 0
        var bestDistance = Double.POSITIVE_INFINITY
        var flip = false
        remaining.forEachIndexed { index, segment ->
            val start = haversine(tail.lat, tail.lon, segment.first().lat, segment.first().lon)
            val end = haversine(tail.lat, tail.lon, segment.last().lat, segment.last().lon)
            val distance = min(start, end)
            if (distance < bestDistance) {
                bestDistance = distance
                bestIndex = index
                flip = end < start
            }
        }
        // A gap larger than 2 km means the rest belongs to a detached branch.
        if (bestDistance > 2000) break
        val next = remaining.removeAt(bestIndex)
        path.addAll(if (flip) next.reversed() else next)
    }
    return path
}

private suspend fun overpass(query: String): OverpassResponse {
    var lastError: Exception? = null
    for (endpoint in listOf(OVERPASS_PRIMARY, OVERPASS_FALLBACK)) {
        try {
            val response =
                HttpClientProvider.client.submitForm(
                    url = endpoint,
                    formParameters = Parameters.build { append("data", query) },
                )
            if (!response.status.isSuccess()) error("Overpass ${response.status.value}")
            return json.decodeFromString<OverpassResponse>(response.bodyAsText())
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            lastError = e
        }
    }
    throw lastError ?: Exception("Overpass unavailable")
}

/** Signposted cycle routes whose ways pass near the given centre. */
suspend fun findNearbyRoutes(
    center: LatLon,
    radiusM: Double,
): List<DiscoveredRoute> {
    val radius = radiusM.coerceIn(2000.0, 40000.0).roundToInt()
    val query =
        """
        [out:json][timeout:40];
        relation["route"="bicycle"]["name"](around:$radius,${"%.5f".format(center.lat)},${"%.5f".format(center.lon)});
        out geom 40;
        """.trimIndent()
    val data = overpass(query)

    val routes = mutableListOf<DiscoveredRoute>()
    for (element in data.elements) {
        if (element.type != "relation") continue
        val path = stitch(element.members)
        if (path.size < 2) continue
        val distanceM = pathLengthM(path)
        // Skip tiny connector segments — they aren't rides.
        if (distanceM < 5000) continue
        val tags = element.tags ?: emptyMap()
        val network = tags["network"] ?: ""
        routes.add(
            DiscoveredRoute(
                id = "osm-${element.id}",
                name = tags["name"] ?: "Unnamed cycle route",
                kind = DiscoveredRouteKind.OSM,
                subtitle = NETWORK_LABEL[network] ?: tags["operator"] ?: "Signposted cycle route",
                path = path,
                distanceM = distanceM,
                ascentM = 0.0,
            ),
        )
    }
    return routes.sortedByDescending { it.distanceM }.take(12)
}

private fun offset(
    center: LatLon,
    bearingDeg: Double,
    distanceM: Double,
): LatLon {
    val earthRadiusM = 6_371_000.0
    val bearing = toRad(bearingDeg)
    val lat1 = toRad(center.lat)
    val lon1 = toRad(center.lon)
    val angular = distanceM / earthRadiusM
    val lat2 = asin(sin(lat1) * cos(angular) + cos(lat1) * sin(angular) * cos(bearing))
    val lon2 = lon1 + atan2(sin(bearing) * sin(angular) * cos(lat1), cos(angular) - sin(lat1) * sin(lat2))
    return LatLon(toDeg(lat2), toDeg(lon2))
}

private fun compass(bearingDeg: Double): String {
    val names = listOf("north", "north-east", "east", "south-east", "south", "south-west", "west", "north-west")
    return names[Math.round((bearingDeg % 360) / 45).toInt() % 8]
}

/**
 * Loop rides of roughly the requested length, starting and finishing at
 * [start]. Each loop heads out on a different bearing. Unlike the web
 * version there's no explicit AbortSignal plumbing — Kotlin's structured
 * concurrency already cancels the in-flight `async` calls below for free
 * when the caller's own coroutine (e.g. a ViewModel's viewModelScope) is
 * cancelled, which is the same effect discover.ts's `signal` param exists
 * to get on the web.
 */
suspend fun generateLoops(
    start: LatLon,
    targetM: Double,
    count: Int = 3,
): List<DiscoveredRoute> =
    coroutineScope {
        // Triangle-ish loop: three waypoints on a circle sized so the ridden
        // distance lands near the target.
        val radius = (targetM / (2 * Math.PI)) * 0.9
        val bearings = List(count) { index -> (360.0 / count) * index + 20 }

        val deferred =
            bearings.mapIndexed { index, bearing ->
                async {
                    val waypoints =
                        listOf(
                            start,
                            offset(start, bearing, radius),
                            offset(start, bearing + 120, radius),
                            offset(start, bearing + 240, radius),
                            start,
                        )
                    // BRouter first (its geometry carries elevation, unlike
                    // OSRM's) with OSRM as the automatic fallback fetchRoute
                    // already provides — same resilience as /plan, just
                    // without a profile choice to honor here.
                    val routed = fetchRoute(waypoints, BikeProfile.TREKKING, preferBrouter = true)
                    // If both routers are down, fetchRoute would otherwise
                    // return a straight-line polygon instead of throwing —
                    // reject that explicitly so this bearing is dropped
                    // below, rather than showing a fake "loop" that ignores
                    // roads.
                    if (!routed.routed) error("No route found for this bearing")
                    val (ascentM, _) = computeAscentDescent(toRidePoints(routed.path))
                    DiscoveredRoute(
                        id = "loop-$index",
                        name = "${Math.round(routed.distanceM / 1000)} km loop · ${compass(bearing)}",
                        kind = DiscoveredRouteKind.LOOP,
                        subtitle = "Generated loop from your start point",
                        path = routed.path,
                        distanceM = routed.distanceM,
                        ascentM = ascentM,
                    )
                }
            }

        deferred.mapNotNull { runCatching { it.await() }.getOrNull() }
    }
